File: android/src/main/kotlin/com/techlabs/flutter/flutter_techlabs_library/lib/model/ServiceInfo.kt
package com.techlabs.flutter.flutter_techlabs_library.lib.model

import org.json.JSONArray
import org.json.JSONObject

/**
 * S3 서비스 정보 JSON 최상위 모델.
 */
data class ServiceInfo(
    val domain: String,
    val maintenance: String,
    val message: String,
    val title: String,
    val appAction: String,
    val appVerInfo: List<AppVerInfo>
) {
    val isUnderMaintenance: Boolean
        get() = maintenance == "Y"

    val androidAppVerInfo: AppVerInfo?
        get() = appVerInfo.firstOrNull { it.device == "android" }

    val iosAppVerInfo: AppVerInfo?
        get() = appVerInfo.firstOrNull { it.device == "ios" }

    companion object {
        fun fromJsonString(json: String): ServiceInfo {
            val jsonObject = JSONObject(json)
            val appVerInfoArray: JSONArray = jsonObject.optJSONArray("appVerInfo") ?: JSONArray()
            val appVerInfoList = (0 until appVerInfoArray.length()).map { i ->
                AppVerInfo.fromJsonObject(appVerInfoArray.getJSONObject(i))
            }
            return ServiceInfo(
                domain = jsonObject.optString("domain", ""),
                maintenance = jsonObject.optString("maintenance", "N"),
                message = jsonObject.optString("message", ""),
                title = jsonObject.optString("title", ""),
                appAction = jsonObject.optString("appAction", ""),
                appVerInfo = appVerInfoList
            )
        }
    }
}

File: lib/src/models/service_info.dart
import 'app_ver_info.dart';

class ServiceInfo {
  final String domain;
  final String maintenance;
  final String message;
  final String title;
  final String appAction;
  final List<AppVerInfo> appVerInfo;

  const ServiceInfo({
    required this.domain,
    required this.maintenance,
    required this.message,
    required this.title,
    required this.appAction,
    required this.appVerInfo,
  });

  bool get isUnderMaintenance => maintenance == 'Y';

  AppVerInfo? get androidAppVerInfo =>
      appVerInfo.where((e) => e.device == 'android').firstOrNull;

  AppVerInfo? get iosAppVerInfo =>
      appVerInfo.where((e) => e.device == 'ios').firstOrNull;

  factory ServiceInfo.fromMap(Map<Object?, Object?> map) {
    final rawList = map['appVerInfo'];
    final List<AppVerInfo> verInfoList;
    if (rawList is List) {
      verInfoList = rawList
          .whereType<Map<Object?, Object?>>()
          .map(AppVerInfo.fromMap)
          .toList();
    } else {
      verInfoList = [];
    }

    return ServiceInfo(
      domain: (map['domain'] as String?) ?? '',
      maintenance: (map['maintenance'] as String?) ?? 'N',
      message: (map['message'] as String?) ?? '',
      title: (map['title'] as String?) ?? '',
      appAction: (map['appAction'] as String?) ?? '',
      appVerInfo: verInfoList,
    );
  }

  Map<String, Object?> toMap() {
    return {
      'domain': domain,
      'maintenance': maintenance,
      'message': message,
      'title': title,
      'appAction': appAction,
      'appVerInfo': appVerInfo.map((e) => e.toMap()).toList(),
    };
  }
}
